firmware/libraries/fast_fifo/fifo.c

```diff
@@ -61,15 +61,16 @@ int fifo_peek(fifo_t *fifo, uint8_t *dest, uint32_t length)
   return length;
 }
 
-void fifo_drop(fifo_t *fifo, uint32_t length)
+int fifo_drop(fifo_t *fifo, uint32_t length)
 {
-  fifo->read_index = (fifo->read_index + length) & (fifo->size - 1);
-  if (length > fifo->bytes_used) {
-    fifo->bytes_used = 0;
-  }
-  else {
-    fifo->bytes_used = fifo->bytes_used - length;
+  if (fifo->bytes_used < length) {
+    return 0;
   }
+
+  fifo->read_index = (fifo->read_index + length) & (fifo->size - 1);
```

Comment:
Can you explain this line, I don't quite get the logic.
For example if index is 4, length is 1, and size is 6:
(4 + 1 = 5) & (6 - 1 = 5) = 5 so new read_index = 5, but if we're dropping a byte why do we change the index from 4 to 5 instead of 4 to 3

Reply:
for this line to work for any arbitrary fifo, the size of the fifo must be a power of 2. 

but since we are dropping a byte, we simply increment the read_index by 1 so in this case going from 4 to 5 makes sense.

before drop:

0 1 2 3 4 5.       ---> indeces
----------- 
3 5 6 9 0 1        ---> data
          ^ ^
           |  |
           |  - read index
           ---- write index  
           
after drop of length 1:

0 1 2 3 4 5.       ---> indeces
----------- 
3 5 6 9 0 1        ---> data
          ^    ^
           |     |
           |     - read index
           ---- write index

Reply:
Oh okay yeah that makes sense, I was confusing myself with lifo. Thanks for the explanation